src/train.py

```diff
@@ -1,118 +1,317 @@
         policy network parameters, CRL critic parameters, their respective optimizer states, alpha parameter,
         and normalization parameters.
     """
-    key_policy, key_sa_enc, key_g_enc = jax.random.split(key, 3)
-    log_alpha = jnp.asarray(0.0, dtype=jnp.float32)
-    alpha_optimizer_state = alpha_optimizer.init(log_alpha)
-
-    c = jnp.asarray(0.0, dtype=jnp.float32)
-
-    policy_params = crl_network.policy_network.init(key_policy)
-    policy_optimizer_state = policy_optimizer.init(policy_params)
-
-    normalizer_params = running_statistics.init_state(specs.Array((obs_size,), jnp.dtype("float32")))
-
-    sa_encoder_params = crl_network.sa_encoder.init(key_sa_enc)
-    g_encoder_params = crl_network.g_encoder.init(key_g_enc)
-    crl_critic_params = {"sa_encoder": sa_encoder_params, "g_encoder": g_encoder_params, "c": c}
-    crl_critic_state = crl_critics_optimizer.init(crl_critic_params)
-
-    training_state = TrainingState(
-        policy_optimizer_state=policy_optimizer_state,
-        policy_params=policy_params,
-        gradient_steps=jnp.zeros(()),
-        env_steps=jnp.zeros(()),
-        alpha_optimizer_state=alpha_optimizer_state,
-        alpha_params=log_alpha,
-        normalizer_params=normalizer_params,
-        crl_critic_optimizer_state=crl_critic_state,
-        crl_critic_params=crl_critic_params,
-    )
-    return jax.device_put_replicated(training_state, jax.local_devices()[:local_devices_to_use])
-
-
-def actor_step(
-    env: Env,
-    env_state: State,
-    policy: Policy,
-    key: PRNGKey,
-    extra_fields: Sequence[str] = (),
-) -> Tuple[State, Transition]:
+    actor_key, sa_key, g_key = jax.random.split(key, 3)
+    
+    # Actor and entropy coefficient
+    actor_params = actor.init(actor_key, jnp.ones([1, state_dim + goal_dim]))
+    actor_state = TrainState.create(apply_fn=actor.apply, params=actor_params, tx=optax.adam(learning_rate=actor_lr))
+    log_alpha = {"log_alpha": jnp.array(0.0)}
+    alpha_state = TrainState.create(apply_fn=None, params=log_alpha, tx=optax.adam(learning_rate=alpha_lr))
+
+    # Critic
+    sa_encoder_params = sa_encoder.init(sa_key, jnp.ones([1, state_dim + action_dim]))
+    g_encoder_params = g_encoder.init(g_key, jnp.ones([1, goal_dim]))
+    critic_params = {"sa_encoder": sa_encoder_params, "g_encoder": g_encoder_params}
+    critic_state = TrainState.create(apply_fn=None, params=critic_params, tx=optax.adam(learning_rate=critic_lr))
+
+    # Put everything together into TrainingState
+    training_state = TrainingState(env_steps=jnp.zeros(()), gradient_steps=jnp.zeros(()), 
+                                   actor_state=actor_state, critic_state=critic_state, alpha_state=alpha_state)
+    training_state = jax.device_put_replicated(training_state, jax.local_devices()[:num_local_devices_to_use])
+    return training_state
+
+# Problem: we have made this a single file, but the abstraction is bad. Users will have a difficult time understanding the
```

Comment:
Later perhaps we can encapsulate some logic into `Agent` class (including `energy`, `log_softmax`  and maybe loss functions? )

Reply:
Noted, I like trying to separate it out but want to make sure the abstraction is reasonably tight. If we call it Agent, while this file is Train, is it clear to users which functionality should be in which file, and does it make it easier to understand rather than needing to learn what functionality is where? For example, I wonder about loss and energy functions; we could put energy functions in Agent but loss feels like it should go in Train, but then separating out loss and energy might be weird.

No clear ideas on this front from me, just noting some thoughts.